### PR TITLE
ANDROID: Switch to Gradle

### PR DESCRIPTION
**Pull request by peterkohaut**
Switch target Android to 29 as it is required by Google Play Store
Switch minimum Android to 16 (4.1) as it is the lowest supported version by NDK
Switch to NDK r21
Removed MIPS and armeabi as they are no longer supported
Renamed Android build names into official ones
Added support for native debugging in Android Studio

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -1942,851 +1943,526 @@ webos)
 	;;
 esac
 
+
 #
-# Determine the C++ compiler
+# Check the CPU architecture
 #
-echo_n "Looking for C++ compiler... "
```
Comment (peterkohaut) [inline configure]: this mostly moving host detection before compiler detection

**Comment by peterkohaut** [comment]
also it works only on linux as path to compiler has hardcoded "linux" in path, this will be fixed

File: configure (hunk 1 above)
Comment (ccawley2011) [reply configure]: In that case, can you split this into two separate commits so that the changes made are easier to understand?

**Comment by ccawley2011** [comment]
> Switch to NDK r20

Is there anything in particular we need from this NDK version? From what I understand from [here](https://android-developers.googleblog.com/2017/03/introducing-android-native-development.html), it's possible to use Gradle with NDK r14, which should allow us to continue to target Android 2.3 as we did previously.

**Comment by peterkohaut** [comment]
> > Switch to NDK r20
> 
> Is there anything in particular we need from this NDK version? From what I understand from [here](https://android-developers.googleblog.com/2017/03/introducing-android-native-development.html), it's possible to use Gradle with NDK r14, which should allow us to continue to target Android 2.3 as we did previously.

Few reasons:

- r19 introduced a new way how it integrates with custom build systems - https://developer.android.com/ndk/guides/other_build_systems and there is no difference in android support between r20  and r19. 
- there is no need for ndk-build at all
- its much easier to start with development on android if sdk and ndk is up-to-date (btw gradle project is importable to android studio without issues)
- google marks every ndk older than r20 as obsolete
- according to https://developer.android.com/about/dashboards/  everything before android v16 has 0.6% distribution and v16 is more than 7 years old

File: configure (hunk 1 above)
Comment (sev-) [reply configure]: I second that, this particular commit is way too big and not clear at all what is going on.

**Comment by sev-** [comment]
Could you please explain how to use this? Does it mean that somehow I could put the project into the Android Studio now?

**Comment by peterkohaut** [comment]
> Could you please explain how to use this? Does it mean that somehow I could put the project into the Android Studio now?

Yes, but only the java part works as expected as the native library is already compiled.
After building the android port you need to open build_android/build.tmp as a project in Android Studio.
To get debugging of native code to work, you need to switch "Debug type" to "Dual" in project configuration

File: configure (hunk 1 above)
Comment (peterkohaut) [reply configure]: I split the change to separate commit, and this change is only doing that - moving a huge chunk of code around, sadly diff is too intelligent and it's displaying it like this

Comment (peterkohaut) [reply configure]: just moving everything between
```
#
# Check the CPU architecture
#
```
and 
```
#
# Backend related stuff
#
```

just before 
```
#
# Determine the C++ compiler
#
```

**Comment by peterkohaut** [comment]
I made patch smaller by not requiring so many changes in ./configure. Would be nice to get this in :)

**Comment by sev-** [comment]
Then edit the PR description? :)

File: configure (hunk 1 above)
Comment (peterkohaut) [reply configure]: this part was removed from this change

**Comment by sev-** [comment]
Merging this. Could you please update Wiki with these build steps?

**Pull request merged**